File: src/main/node/direction.js
"use strict";

const Direction = Object.freeze({"LEFT":0,"UP":1,"RIGHT":2,"DOWN":3});
const positions = Object.keys(Direction).length;

function facing(facing) {
  switch (facing) {
    case "LEFT":
      return Direction.LEFT;
    case "RIGHT":
      return Direction.RIGHT;
    case "UP":
      return Direction.UP;
    case "DOWN":
      return Direction.DOWN;
    default:
      return -1;
  }
}

function nextLocation(loc, direction) {
  switch (direction) {
    case Direction.LEFT:
      return {x: loc.x-1, y:loc.y};
    case Direction.RIGHT:
      return {x: loc.x+1, y:loc.y};
    case Direction.UP:
      return {x: loc.x, y:loc.y-1};
    case Direction.DOWN:
      return {x: loc.x, y:loc.y+1};
  }
  return null;
}

function turnLeft(facing) {
  return (facing + 3) % positions;
}

function turnRight(facing) {
  return (facing +1) % positions;
}

function opposite(facing) {
  return (facing +2) % positions;
}

function avoid(ours, theirs) {
  if (ours.x < theirs.x) {
    return Direction.LEFT;
  } else if (ours.x > theirs.x) {
    return Direction.RIGHT;
  }
  if (ours.y < theirs.y) {
    return Direction.UP;
  } else {
    return Direction.DOWN;
  }
}

module.exports = {
  Direction: Direction,
  facing: facing,
  nextLocation: nextLocation,
  turnLeft: turnLeft,
  turnRight: turnRight,
  opposite: opposite,
  avoid: avoid
}

File: src/main/python/direction.py
from enum import IntEnum


class Direction(IntEnum):
    LEFT = 0
    UP = 1
    RIGHT = 2
    DOWN = 3


def next_location(coords, direction):
    return {
        Direction.LEFT: lambda coords: ((coords[0] - 1),
                                        coords[1]),
        Direction.RIGHT: lambda coords: ((coords[0] + 1),
                                         coords[1]),
        Direction.UP: lambda coords: (coords[0],
                                      (coords[1] - 1)),
        Direction.DOWN: lambda coords: (coords[0],
                                        (coords[1] + 1))
        }[direction](coords)


def avoid(our, their):
    if our[0] < their[0]:
        return Direction.LEFT
    elif our[0] > their[0]:
        return Direction.RIGHT
    elif our[1] < their[1]:
        return Direction.DOWN
    else:
        return Direction.UP


def left(direction):
    return ((direction + 3) % 4)


def right(direction):
    return ((direction + 1) % 4)


def opposite(direction):
    return ((direction + 2) % 4)


def wrap(coords, width, height):
    return ((coords[0] + width) % width, (coords[1] + height) % height)
